thorlcr/activities/msort/thmsortslave.cpp

```diff
@@ -168,30 +167,28 @@ class MSortSlaveActivity : public CSlaveActivity
     virtual void reset() override
     {
         PARENT::reset();
-        if (sorter) return; // JCSMORE loop - shouldn't have to recreate sorter between loop iterations
-        sorter.setown(CreateThorSorter(this, server, &queryJobChannel().queryJobComm(), mpTagRPC));
-    }
-    virtual void kill() override
-    {
         {
+            // JCSMORE - it would be better if sorter was reused, but it's non-trivial to do so
             CriticalBlock block(statsCs);
-            mergeStats(stats, sorter, spillStatistics);
-            sorter.clear();
+            // fold the outgoing 'sorter' stats into inactiveStats, which will be merged during CSlaveActivity::serializeStats
+            mergeStats(inactiveStats, sorter, spillStatistics);
+            someInactiveStats = true;
+            sorter.setown(CreateThorSorter(this, server, &queryJobChannel().queryJobComm(), mpTagRPC));
         }
+    }
+    virtual void kill() override
+    {
         if (portbase)
         {
             queryJobChannel().freePort(portbase, NUMSLAVEPORTS);
             portbase = 0;
         }
         PARENT::kill();
     }
-    virtual void serializeStats(MemoryBuffer &mb) override
+    // NB: always called within statsCs crit
+    virtual void gatherActiveStats(CRuntimeStatisticCollection &activeStats) const
     {
-        {
-            CriticalBlock block(statsCs);
-            mergeStats(stats, sorter, spillStatistics);
-        }
-        PARENT::serializeStats(mb);    
+        mergeStats(activeStats, sorter, spillStatistics);
```

Comment:
Does this need protecting with statsCs since sorter may be being replaced inside kill().

Reply:
all calls to gatherActiveStats are called within statsCS (comment on line 188).